yt_idefix/data_structures.py

```diff
@@ -724,3 +793,272 @@ def _validate_units_override_keys(cls, units_override):
+        # on-disk units.  These are the defaults, but if they are listed
+        # in the HDF5 attributes for a file, which is loaded first, then those are
+        # used instead.
+        #
+        from .definitions import pluto_def_constants
+        from .misc import remove_comments
+
+        length_unit = None
+        mass_unit = None
+        velocity_unit = None
+        density_unit = None
+        magnetic_unit = None
+
+        if os.path.exists(self._default_definitions_header):
+            with open(self._default_definitions_header) as deftxt:
+                lines = deftxt.readlines()
+                for line in lines:
+                    line = remove_comments(line)
+                    if "UNIT_LENGTH" in line:
+                        length_unit = line.split()[-1]
+                    if "UNIT_DENSITY" in line:
+                        density_unit = line.split()[-1]
+                    if "UNIT_VELOCITY" in line:
+                        velocity_unit = line.split()[-1]
+
+            constant_names = list(pluto_def_constants.keys())
+            # if definitions.h uses inputParamaeter then these values need to be read from pluto.ini
+            g_inputParam = {}
+            if (
+                ("g_inputParam" in length_unit)
+                or ("g_inputParam" in density_unit)
+                or ("g_inputParam" in velocity_unit)
+            ):
+                ini_file = os.path.join(
+                    os.path.dirname(self.parameter_filename), "pluto.ini"
+                )
+                if not (os.path.exists(ini_file)):
+                    raise RuntimeError(
+                        "pluto.ini file is needed for unit conversion but is missing!"
+                    )
+                with open(ini_file) as ini:
+                    lines = ini.readlines()
+                    for _pos, line in enumerate(lines):
+                        if "[Parameters]" in line:
+                            break
+                    tmp = 0
+                    while tmp == 0:
+                        _pos += 1
+                        tmp = len(lines[_pos].split())
+                    for i in range(_pos, _pos + self.inputParamCount):
+                        txt = lines[i].split()
+                        g_inputParam[txt[0]] = float(txt[1])
+            inputParam_names = list(g_inputParam.keys())
+
+            if length_unit:
+                try:
+                    length_unit = float(length_unit)
+                except ValueError:
+                    for name in constant_names:
+                        length_unit = length_unit.replace(
+                            name, f'pluto_def_constants["{name}"]'
+                        )
+                    for name in inputParam_names:
+                        length_unit = length_unit.replace(name, f'"{name}"')
+                    length_unit = length_unit.replace("sqrt", "np.sqrt")
+                    length_unit = length_unit.replace("log", "np.log")
```

Comment:
I'm not sure if `log` will be used in the unit definitions, could you show me an example? If it's possile, I will add it in my next PR too.

Reply:
Yes, for example one of our code we were using M_star as a parameter and using the M_star-M_halo relation, which involves a logarithmic relation, we were using M_halo as unit mass by setting UNIT_DENSITY. This is because the natural Mass sizes to scale would be for gas
This is not so common. However, the more common would be power (many astrophysical quantities are expressed in log10) which I also now added.

Reply:
Thanks, I got it!